### PR TITLE
[frontend-api] registration after order refactoring

### PR DESCRIPTION
**Pull request by vitek-rostislav**
#### Description, the reason for the PR
When a customer decided to register after he finished the order, optional delivery address data entered in the order were missing from his newly created profile. This is fixed now. The code was refactored and the new `registerByOrderMutation` was separated from `registerMutation`

#### Fixes issues
... 

#### Check the appropriate checkboxes below, please

- [x] [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/) 
- [ ] New feature 
- [x] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new registration method via orders using `RegistrationByOrderInput`.
	- Added functionality to merge product lists upon registration.
	- Enhanced user registration process with improved validation and error handling.

- **Bug Fixes**
	- Streamlined logic for handling registration data and order URL hash.

- **Documentation**
	- Updated GraphQL schema documentation to reflect new input types and mutations.

- **Translations**
	- Revised validation messages for clarity and consistency in both Czech and English.

----
:globe_with_meridians: Live Preview:
  - https://rv-delivery-fields-registration.odin.shopsys.cloud
  - https://cz.rv-delivery-fields-registration.odin.shopsys.cloud

### CODE REVIEW

**Comment by grossmannmartin** [comment]
@coderabbitai review

File: packages/frontend-api/src/Model/Customer/User/CustomerUserUpdateDataFactory.php [hunk 1]
```diff
@@ -57,6 +58,16 @@ public function createFromCustomerUser(CustomerUser $customerUser): CustomerUser
         return $this->customerUserUpdateDataFactory->createFromCustomerUser($customerUser);
     }
 
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\Order $order
+     * @param string $password
+     * @return \Shopsys\FrameworkBundle\Model\Customer\User\CustomerUserUpdateData
+     */
+    public function createFromOrder(Order $order, string $password): CustomerUserUpdateData
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: do we need the duplicated method on framework and frontend-api? 

This raises a few questions I don't want to find answers to, like:
- should they be always the same? 
- what method should I use in my code?

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php [hunk 2]
```diff
@@ -180,4 +199,43 @@ protected function getAmendedDeliveryAddressDataByOrder(Order $order, ?DeliveryA
...
+     * @return \Shopsys\FrameworkBundle\Model\Customer\User\CustomerUserData
+     */
+    protected function getCustomerUserDataByOrder(Order $order, string $password): CustomerUserData
+    {
+        $customerUserData = $this->customerUserDataFactory->createForDomainId($order->getDomainId());
+        $customerUserData->firstName = $order->getFirstName();
+        $customerUserData->lastName = $order->getLastName();
+        $customerUserData->telephone = $order->getTelephone();
+        $customerUserData->email = $order->getEmail();
+        $customerUserData->createdAt = new DateTime();
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: _nitpick_ I would not set the value and let the entity itself handle the correct createdAt value

File: packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php [hunk 3]
```diff
@@ -58,6 +67,32 @@ public function register(RegistrationData $registrationData): CustomerUser
...
+     */
+    public function registerByOrder(string $orderUrlHash, string $password): CustomerUser
+    {
+        $order = $this->orderFacade->getByUrlHashAndDomain($orderUrlHash, $this->domain->getId());
+
+        if ($order->getCustomerUser() !== null) {
+            throw new OrderCannotBePairedException('Order is owned by another customer.');
+        }
+
+        if ($order->getCreatedAt()->getTimestamp() < (time() - self::ONE_HOUR_REGISTRATION_WINDOW)) {
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php]: what about this?
```
$order->getCreatedAt() < new DateTime('-1 hour') 
```

or in seconds using the constant if you want to keep the constant

File: packages/frontend-api/src/Resources/config/graphql-types/ModelType/Customer/CustomerUser/Input/RegistrationByOrderInputDecorator.types.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+RegistrationByOrderInputDecorator:
+    type: input-object
+    decorator: true
+    config:
+        fields:
+            orderUrlHash:
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Resources/config/graphql-types/ModelType/Customer/CustomerUser/Input/RegistrationByOrderInputDecorator.types.yaml:6]: should we use urlhash or uuid? in `AddOrderItemsToCartInputDecorator` we use UUID... I'm not sure where is the line of using one or the other

File: packages/framework/src/Model/Newsletter/NewsletterFacade.php [hunk 5]
```diff
@@ -108,4 +109,13 @@ public function deleteSubscribedEmailIfExists(string $email, int $domainId): voi
...
         $this->delete($newsletterSubscriber);
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Customer\User\CustomerUser $customerUser
+     * @return bool
+     */
+    public function isSubscribed(CustomerUser $customerUser): bool
+    {
+        return $this->findNewsletterSubscriberByEmailAndDomainId($customerUser->getEmail(), $customerUser->getDomainId()) !== null;
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Newsletter/NewsletterFacade.php:119]: can we talk about this commit please? 

We're replacing a simple boolean on already hydrated entity with potentialy expensive database query

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/frontend-api/src/Resources/translations/validators.cs.po [hunk 6]
```diff
@@ -127,8 +127,11 @@ msgstr "Zadejte staré heslo"
 msgid "Please enter only numbers and the + character"
 msgstr "Prosím, zadávejte pouze čísla a znak +"
 
-msgid "Please enter password"
-msgstr "Vyplňte prosím heslo"
+msgid "Please enter only numbers and the + sign"
+msgstr "Prosím vyplňte čísla a znak +"
+
+msgid "Please enter order URL hash"
+msgstr "Vyplňte prosím URl hash objednávky"
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Resources/translations/validators.cs.po]: 👍 @vitek-rostislav

File: packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php [hunk 7]
```diff
@@ -4,26 +4,35 @@
...
 use Shopsys\FrameworkBundle\Model\Customer\User\CustomerUser;
 use Shopsys\FrameworkBundle\Model\Customer\User\CustomerUserFacade;
 use Shopsys\FrameworkBundle\Model\Customer\User\CustomerUserUpdateData;
 use Shopsys\FrameworkBundle\Model\Newsletter\NewsletterFacade;
+use Shopsys\FrameworkBundle\Model\Order\OrderFacade;
+use Shopsys\FrontendApiBundle\Model\Order\Exception\OrderCannotBePairedException;
 
 class RegistrationFacade
 {
+    protected const int ONE_HOUR_REGISTRATION_WINDOW = 3600;
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php]: @coderabbitai they are in php 8.3. We already talked about it.

File: project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $order = $this->getReference(OrderDataFixture::ORDER_PREFIX . '7', Order::class);
+
+        $response = $this->getResponseContentForGql(__DIR__ . '/../../_graphql/mutation/RegistrationByOrderMutation.graphql', [
+            'orderUrlHash' => $order->getUrlHash(),
+            'password' => 'user123',
+        ]);
+
+        $this->assertOrderCannotBePairedUserErrorIsReturned($response);
+    }
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php]: @vitek-rostislav  This seems interesting, but it's not necessary to me. If you dont want to change it, just close this comment :)

File: project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+        $graphQlType = 'RegisterByOrder';
+        $this->assertResponseContainsArrayOfDataForGraphQlType($response, $graphQlType);
+        $responseData = $this->getResponseDataForGraphQlType($response, $graphQlType);
+
+        $this->assertArrayHasKey('tokens', $responseData);
+        $this->assertIsString($responseData['tokens']['accessToken']);
+
+        $this->assertArrayHasKey('tokens', $responseData);
+        $this->assertIsString($responseData['tokens']['refreshToken']);
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php]: 👍 @vitek-rostislav

File: project-base/storefront/components/Pages/OrderConfirmation/RegistrationAfterOrder.tsx [hunk 10]
```diff
@@ -27,14 +26,11 @@ export const RegistrationAfterOrder: FC = () => {
     const { t } = useTranslation();
     const [formProviderMethods] = useRegistrationAfterOrderForm();
     const formMeta = useRegistrationAfterOrderFormMeta(formProviderMethods);
-    const register = useRegistration();
+    const { registerByOrder } = useRegistration();
     const isInvalidRegistrationRef = useRef(false);
     const { query } = useRouter();
-    const { orderUuid, orderEmail, registrationData } = query as OrderConfirmationUrlQuery;
+    const { orderUuid, orderEmail, orderUrlHash } = query as OrderConfirmationUrlQuery;
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/storefront/components/Pages/OrderConfirmation/RegistrationAfterOrder.tsx:32]: this probably relates to the question whether we should use uuid or urlhash

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php [hunk 11]
```diff
@@ -4,6 +4,7 @@
 
 namespace Shopsys\FrameworkBundle\Model\Customer\User;
 
+use DateTime;
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: Although nice thought, this is not possible without more changes

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php [hunk 12]
```diff
@@ -53,6 +54,30 @@
...
+     * @param \Shopsys\FrameworkBundle\Model\Order\Order $order
+     * @param string $password
+     * @return \Shopsys\FrameworkBundle\Model\Customer\User\CustomerUserUpdateData
+     */
+    public function createFromOrder(Order $order, string $password): CustomerUserUpdateData
+    {
+        $customerUserUpdateData = $this->create();
+
+        $customerUserUpdateData->customerUserData = $this->getCustomerUserDataByOrder($order, $password);
+        $customerUserUpdateData->billingAddressData = $this->fillBillingAddressDataFromOrder($order, $customerUserUpdateData->billingAddressData);
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: 👍 @vitek-rostislav

File: packages/frontend-api/src/Model/Mutation/Customer/User/CustomerUserMutation.php [hunk 13]
```diff
@@ -155,26 +155,22 @@ public function registerMutation(Argument $argument, InputValidator $validator):
...
 
-        return $this->loginResultDataFactory->create(
-            $this->tokensDataFactory->create(
-                $this->tokenFacade->createAccessTokenAsString($customerUser, $deviceId),
-                $this->tokenFacade->createRefreshTokenAsString($customerUser, $deviceId),
-            ),
-            $this->mergeCartFacade->shouldShowCartMergeInfo(),
-            true,
-        );
+        return $this->loginRegisteredCustomerUser($customerUser);
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Mutation/Customer/User/CustomerUserMutation.php:173]: 👍 @vitek-rostislav  the invalid url hash is a valid point

File: packages/frontend-api/src/Resources/translations/validators.cs.po (hunk 6 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/frontend-api/src/Resources/translations/validators.cs.po]: nice catch, fixed - a074d5893c

File: project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php]: love it - 64aca775ab

File: project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply project-base/app/tests/FrontendApiBundle/Functional/Customer/User/RegisterByOrderTest.php]: true, fixed in c5d2eb94bb

File: packages/frontend-api/src/Model/Mutation/Customer/User/CustomerUserMutation.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Mutation/Customer/User/CustomerUserMutation.php:173]: good catch, fixed in 5a345790c4

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: makes sense, fixed in ff7501d220

File: packages/frontend-api/src/Model/Customer/User/CustomerUserUpdateDataFactory.php (hunk 1 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: good point, I was just lazy and did not want to wire both `CustomerUserUpdateDataFactory` from the `frontend-api` and `framework` packages in `RegistrationFacade`. However, fixed in 2d8092a03a

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php (hunk 2 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: makes sense, 321c3b92cf

File: project-base/storefront/components/Pages/OrderConfirmation/RegistrationAfterOrder.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply project-base/storefront/components/Pages/OrderConfirmation/RegistrationAfterOrder.tsx:32]: discussed, we are keeping urlHash as it is a bit more secure

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php (hunk 2 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: makes sense, 321c3b92cf

File: packages/frontend-api/src/Resources/config/graphql-types/ModelType/Customer/CustomerUser/Input/RegistrationByOrderInputDecorator.types.yaml (hunk 4 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Resources/config/graphql-types/ModelType/Customer/CustomerUser/Input/RegistrationByOrderInputDecorator.types.yaml:6]: discussed, urlHash is a bit more secure

File: packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php (hunk 3 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Customer/User/RegistrationFacade.php]: :ok_hand: 43577535a6

File: packages/framework/src/Model/Newsletter/NewsletterFacade.php (hunk 5 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Newsletter/NewsletterFacade.php:119]: discussed, keeping as is, I will create an issue for optimizing the current customer queries on the SF

**Review by grossmannmartin — APPROVED** [review]

**Pull request merged**